src/pip/_internal/wheel.py

```diff
@@ -1036,14 +1037,22 @@ def build(
         # type: (...) -> List[InstallRequirement]
         """Build wheels.
 
-        :param unpack: If True, replace the sdist we built from with the
-            newly built wheel, in preparation for installation.
+        :param autobuilding: If True, replace the sdist we built from with the
+            unpacked newly built wheel, in preparation for installation.
         :return: True if all the wheels built correctly.
         """
+        # pip install uses autobuilding=True.
+        # pip install never provides a _wheel_dir.
+        # pip wheel uses autobuilding=False.
+        # pip wheel always provides a _wheel_dir (via the preparer).
+        assert (
+            (autobuilding and not self._wheel_dir) or
+            (not autobuilding and self._wheel_dir)
+        )
+
         buildset = []
         format_control = self.finder.format_control
-        # Whether a cache directory is available for autobuilding=True.
-        cache_available = bool(self._wheel_dir or self.wheel_cache.cache_dir)
```

Comment:
Since this was working before without first checking that `self.wheel_cache` is truthy, does this mean we can add the following to the above asserts?

```python
if autobuilding:
    assert self.wheel_cache
```

(or add `and self.wheel_cache` to the first expression of your existing assert)

Reply:
The failing test was due to this mock providing no wheelcache:
https://github.com/pypa/pip/blob/76a89542a8ea8b9ebbfe393f23ccbe764c727d3e/tests/unit/test_wheel.py#L700

Reply:
So I fixed the test instead.